torch/csrc/inductor/aot_runtime/model.h

```diff
@@ -40,9 +41,77 @@
 namespace torch {
 namespace aot_inductor {
 
-using RAIIAtenTensorHandle = std::shared_ptr<AtenTensorOpaque>;
+// UniqueAtenTensorHandle is used to manage input, output, and intermediate
+// buffers
+class UniqueAtenTensorHandle {
+ public:
+  UniqueAtenTensorHandle() = delete;
+  UniqueAtenTensorHandle(const UniqueAtenTensorHandle& other) = delete;
+  UniqueAtenTensorHandle& operator=(const UniqueAtenTensorHandle& other) =
+      delete;
+
+  // Steal the ownership from another UniqueAtenTensorHandle using std::move
+  UniqueAtenTensorHandle(UniqueAtenTensorHandle&& other) = default;
+  UniqueAtenTensorHandle& operator=(UniqueAtenTensorHandle&& other) = default;
+
+  // Steal the ownership from raw AtenTensorHandle
+  UniqueAtenTensorHandle(AtenTensorHandle handle)
+      : handle_(handle, [](AtenTensorHandle ptr) {
+          AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_delete_tensor_object(ptr));
+        }) {}
+
+  ~UniqueAtenTensorHandle() {
+    handle_.reset();
+  }
+
+  // Return a raw AtenTensorHandle to be used by aoti_torch functions
+  // Note: this function does NOT transfer the ownership of the handle
+  operator AtenTensorHandle() const {
+    return handle_.get();
+  }
+
+  AtenTensorHandle release() {
+    return handle_.release();
+  }
+
+  AtenTensorHandle get() {
+    return handle_.get();
+  }
+
+  void reset() {
+    handle_.reset();
+  }
+
+ private:
+  std::unique_ptr<AtenTensorOpaque, std::function<void(AtenTensorOpaque*)>>
```

Comment:
micro optimization note: you should be able to do this with a raw function pointer deleter (no need for std::function, which is famously slow)